### PR TITLE
Implement task 2.1

### PR DESCRIPTION
**Pull request by buyfn**

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+router.route('/users')
+    .put((req, res) => {
+        const user = req.body;
+        const id = addUser(user);
+
+        res.json({ id });
+    });
+
+app.use('/', router);
+app.listen(PORT);
```
Comment (titivuk) [inline src/index.js]: app.listen supports 2nd call back parameter where you can add log that server started.

File: src/index.js [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+process.stdout.write(`Server started on port ${PORT}\n`);
```
Comment (titivuk) [inline src/index.js]: use console object to log something instead of process.stdout

File: src/index.js [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+let users = [];
+const PROPERTIES = [
+    'login',
+    'password',
+    'age'
+];
+
+const findUser = (id) => users
+    .filter(user => !user.isDeleted)
+    .find(user => user.id === id);
```
Comment (titivuk) [inline src/index.js]: you can apply `isDeleted` filtering in place.
for example, `find(user => user.id === id && !user.isDeleted)`

File: src/index.js [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+        res.sendStatus(204);
+    })
+    .delete((req, res) => {
+        removeUser(req.params.id);
+
+        res.sendStatus(204);
+    });
+
+router.route('/users')
+    .put((req, res) => {
```
Comment (titivuk) [inline src/index.js]: when you create new object, usually it's `POST`

File: src/index.js [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+    .get((req, res) => {
+        const user = findUser(req.params.id);
+
+        if (user) {
+            res.json(user);
+        } else {
+            res.status(404).json({ message: 'User not found' });
+        }
+    })
+    .post((req, res) => {
```
Comment (titivuk) [inline src/index.js]: when you update object, it can be `PUT` or `PATCH`

File: src/index.js [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+
+        res.sendStatus(204);
+    });
+
+router.route('/users')
+    .put((req, res) => {
+        const user = req.body;
+        const id = addUser(user);
+
+        res.json({ id });
```
Comment (titivuk) [inline src/index.js]: in most cases we return back what we've just created, not just id

File: src/index.js [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+        if (userA.login > userB.login) {
+            return 1;
+        }
+        return 0;
+    })
+    .slice(0, limit);
+
+const router = express.Router();
+
+router.get('/users/suggest', (req, res) => {
```
Comment (titivuk) [inline src/index.js:76]: you dont need separate router, you operate on the same entity - User. Please, apply all routers to single router or app

File: src/index.js [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+    'login',
+    'password',
+    'age'
+];
+
+const findUser = (id) => users
+    .filter(user => !user.isDeleted)
+    .find(user => user.id === id);
+
+const addUser = (user) => {
```
Comment (titivuk) [inline src/index.js:19]: I'm almost sure that login has to be unique even though it's not mentioned in the task

File: src/index.js [hunk 9]
```diff
@@ -0,0 +1,111 @@
+import express from 'express';
+import { v4 as uuidv4 } from 'uuid';
+import _ from 'lodash';
+
+const PORT = Number(process.env.PORT) || 3000;
+const app = express();
```
Comment (titivuk) [inline src/index.js:8]: image
`password` and `isDeleted` accessible through the API. I think these fields should never be returned to the user

File: src/index.js [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+app.use(express.json());
```
Comment (titivuk) [inline src/index.js:9]: also add unhandled exception handler. For now, if exception happens, your API returns HTML

File: src/validation.js [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+
+const errorResponse = (schemaErrors) => {
+    const errors = schemaErrors.map(error => _.pick(error, ['path', 'message']));
+    return {
+        status: 'failed',
+        errors
+    };
+};
+
+export const validateSchema = (s) => {
```
Comment (titivuk) [inline src/validation.js]:
```suggestion
export const validateSchema = (schema) => {
```

avoid non descriptive naming

File: src/index.js (hunk 1 above)
Comment (buyfn) [reply src/index.js]: Moved log to the callback

File: src/index.js (hunk 2 above)
Comment (buyfn) [reply src/index.js]: Updated

File: src/index.js (hunk 3 above)
Comment (buyfn) [reply src/index.js]: Updated

File: src/index.js (hunk 4 above)
Comment (buyfn) [reply src/index.js]: Changed method

File: src/index.js (hunk 5 above)
Comment (buyfn) [reply src/index.js]: Changed method to `PATCH`

File: src/index.js (hunk 6 above)
Comment (buyfn) [reply src/index.js]: Updated to return all user's info.

File: src/index.js (hunk 7 above)
Comment (buyfn) [reply src/index.js:76]: Resolved in chat

File: src/index.js (hunk 8 above)
Comment (buyfn) [reply src/index.js:19]: Decided not to do in chat

File: src/index.js (hunk 9 above)
Comment (buyfn) [reply src/index.js:8]: Removed `password` and `isDeleted` from responses.

File: src/index.js (hunk 10 above)
Comment (buyfn) [reply src/index.js:9]: Added error handler.

File: src/validation.js (hunk 11 above)
Comment (buyfn) [reply src/validation.js]: Updated

**Pull request merged**